cache_flat_mutation_reader.hh

```diff
@@ -275,8 +275,16 @@ future<> cache_flat_mutation_reader::do_fill_buffer(db::timeout_clock::time_poin
             }
         };
         if (!_underlying) {
-            return ensure_underlying(read_context::skip_first_fragment_t::yes, timeout).then([this, timeout] (read_context::created_new_underlying_t){
-                return do_fill_buffer(timeout);
+            return ensure_underlying(read_context::skip_first_fragment_t::no, timeout).then([this, timeout, prepare_underlying = std::move(prepare_underlying)] (read_context::created_new_underlying_t created_new_underlying) mutable {
```

Comment:
I think it would be overall simpler if we put this logic in read_context. We add a flag `read_context::partition_exists_in_underlying()`, and the impl of `ensure_underlying()` and `create_underlying()` would set it. 

Then, this layer doesn't have to deal with the details of whether the underlying is reused or not.

Something like this:
```diff
diff --git a/cache_flat_mutation_reader.hh b/cache_flat_mutation_reader.hh
index 54a578c60b..b11d214536 100644
--- a/cache_flat_mutation_reader.hh
+++ b/cache_flat_mutation_reader.hh
@@ -267,6 +267,9 @@ future<> cache_flat_mutation_reader::do_fill_buffer(db::timeout_clock::time_poin
         }
         _state = state::reading_from_underlying;
         _population_range_starts_before_all_rows = _lower_bound.is_before_all_clustered_rows(*_schema);
+        if (!_read_context->partition_exists()) {
+            return read_from_underlying(timeout);
+        }
         auto end = _next_row_in_range ? position_in_partition(_next_row.position())
                                       : position_in_partition(_upper_bound);
         return _underlying->fast_forward_to(position_range{_lower_bound, std::move(end)}, timeout).then([this, timeout] {
diff --git a/read_context.hh b/read_context.hh
index 29010618cf..7170b7486d 100644
--- a/read_context.hh
+++ b/read_context.hh
@@ -137,6 +137,7 @@ class read_context final : public enable_lw_shared_from_this<read_context> {
     //     are set so that _underlying_flat can be fast forwarded to the right partition.
     //
     autoupdating_underlying_reader _underlying;
+    bool _partition_exists;
     uint64_t _underlying_created = 0;
 
     mutation_source_opt _underlying_snapshot;
@@ -190,13 +191,19 @@ class read_context final : public enable_lw_shared_from_this<read_context> {
     autoupdating_underlying_reader& underlying() { return _underlying; }
     row_cache::phase_type phase() const { return _phase; }
     const dht::decorated_key& key() const { return *_key; }
+    bool partition_exists() const { return _partition_exists; }
     void on_underlying_created() { ++_underlying_created; }
     bool digest_requested() const { return _slice.options.contains<query::partition_slice::option::with_digest>(); }
 public:
     future<> ensure_underlying(db::timeout_clock::time_point timeout) {
         if (_underlying_snapshot) {
-            return create_underlying(true, timeout);
+            return create_underlying(false, timeout).then([this] {
+                return _underlying.underlying()(timeout).then([this](auto&& mfopt) {
+                    _partition_exists = bool(mfopt);
+                });
+            })
         }
+        _partition_exists = true;
         return make_ready_future<>();
     }
 public:

```

Reply:
I agree that the patch is smaller then but I'm not a fan of adding semi-global state when the same information can go through a regular function param/return value path. read_context lifetime is much longer than the call to do_fill_buffer. When we add state to read_context we're leaking complexity from a localised scope of function calls stack/chain to much broader scope of cache_flat_mutation_reader lifetime.

In general, I think we should avoid state and side effects when possible because it (avoiding state and side effects) makes the code cleaner and easier to understand. Having a use case specific logic `L` that requires some input data `I`, I think :
```
I input = shared_object_with_much_longer_lifetime.function_with_limited_side_effects();
<input being returned from 2 levels of stack/function calls>
auto result = L(input);
```
is cleaner/easier to understand than:
```
shared_object_with_much_longer_lifetime.function_with_side_effect_that_applies_L_to_I_and_saves_the_state();
<2 levels of stack/function calls>
auto result = shared_object_with_much_longer_lifetime.get_state_with_result_of_L_applied_to_I();
```

I know that `read_context::ensure_underlying` is not a pure function but adding even more state makes it even less pure. The more pure the function the better in general because it's easier to reason about functions that take input returns output and doesn't have side effects. So I think that it's better to keep `read_context::ensure_underlying` side effects to the minimum. The fact that it sometimes creates the underlying as a side effect is somehow understandable/expected from its name.

So I guess my argument is that `was_underlying_created read_context::ensure_underlying(...)` is easy to understand by just looking at a declaration. While the relation between `void read_context::ensure_underlying(...)` and `bool read_context::partition_exists()` needs to be discovered by reading the implementation so I think it's a less useful abstraction that is more demanding on the reader that tries to understand what's going on.

Reply:
BTW I see your point about exposing whether underlying was reused or not but:
1. figuring out that partition exists or not in underlying is at the moment needed in just one use case so I think it's better to keep the logic where it's needed than pushing it into shared object and extending the scope it lives in.
2. I find it less complex to expose this instead of adding a state to a shared object with much longer lifetime

Reply:
I think there's a hybrid approach possible that avoids both new state and exposition of internals of ensure_underlying. Will post a draft today/tomorrow.

Reply:
Here's how a hybrid approach would work: https://github.com/scylladb/scylla/compare/master...haaawk:fix_cache_flat_mutation_reader_alternative

It encapsulates the logic that checks underlying being in non-existent partition into read_context but at the same time does not introduce more state. @tgrabiec what do you think?

Reply:
> Here's how a hybrid approach would work: [master...haaawk:fix_cache_flat_mutation_reader_alternative](https://github.com/scylladb/scylla/compare/master...haaawk:fix_cache_flat_mutation_reader_alternative)
> 

This hybrid approach looks better than the previous one.

I think there is a problem with it (and with the previous one), that I didn't mention. It's not enough to inhibit f_f_t() at the time we create the underlying. If there is no partition in the underlying, we must not f_f_t() each time we try to read from this partition, even if the underlying reader is ensured for the current partition. So we need to record partition_exists somewhere.

I don't accept the criticism of adding state to read_context. It's true that stateless is in general better than stateful, however, in this case we do need to hold the state. What we need is a cursor over partitions, and part of read_context state is for implementing this cursor. ensure_underlying() makes sure the cursor is valid for the current partition. We already have state for the current partition there (e.g. `read_context::_key`). Adding more state would be in line with this design.

Reply:
> I think there is a problem with it (and with the previous one), that I didn't mention. It's not enough to inhibit f_f_t() at the time we create the underlying. If there is no partition in the underlying, we must not f_f_t() each time we try to read from this partition, even if the underlying reader is ensured for the current partition. So we need to record partition_exists somewhere.

Ok. I get your point. If you have started with that I wouldn't have wasted time fighting the inevitable :)

I will rework the PR.

Reply:
> I don't accept the criticism of adding state to read_context. It's true that stateless is in general better than stateful, however, in this case we do need to hold the state. What we need is a cursor over partitions, and part of read_context state is for implementing this cursor. ensure_underlying() makes sure the cursor is valid for the current partition. We already have state for the current partition there (e.g. `read_context::_key`). Adding more state would be in line with this design.

The criticism was about adding a state to make the code simpler. When the state is needed for correctness then it's a whole different story.